h2o-core/src/main/java/water/parser/Parser.java

```diff
@@ -87,7 +87,10 @@ ParseWriter streamParseZip( final InputStream is, final StreamParseWriter dout,
     StreamParseWriter nextChunk = dout;
     int zidx = bvs.read(null,0,0); // Back-channel read of chunk index
     assert zidx==1;
-    while( is.available() > 0 ) {
+    int streamAvailable = is.available();
+    while(streamAvailable > 0) {
+      parseChunk(cidx++, din, nextChunk);
+      streamAvailable = is.available(); // Can (also!) rollover to the next input chunk
```

Comment:
Hmm, I think it is okay that available() fetches the next chunk, it needs to find out how many bytes are available so it seems almost unavoidable.

From the point of view of this bug:
it is important to keep all the "operations with side effect" (parseChunk, is.available) next to each other (in the source code) and after they are done always check whether the next chunk was fetched or not